### DISCUSSION TITLE
Modify Accounts to enable custom login processes

### DISCUSSION THREAD

**Discussion by jkasevits** [meteor/meteor#11896] [Feature Requests]
To start, a little bit of background. In a couple of apps I need to set up the following login flow:

1. on login page, user enters their email/username and password
2. if multi-factor authentication is not enabled for this particular user account, the user is logged in immediately
3.  if multi-factor authentication is enabled for this user account, then the server checks which particular method the user has enabled (email, SMS, possibly some other method in the future) and displays this list on the login page for the user to choose from
4. on login page user chooses the method they want to use for receiving the one time password (email, SMS)
5. one time password is sent to the user, via the method selected by the user
6. user enters the one time password on the login page and gets logged in

So essentially the client side login page actually consist of 3 views:
1. the first view with email/username and password fields (accounts that do not have MFA enabled only see this)
2. the second view where user chooses the MFA method (if they have enabled more than one)
3. and the third view where the user enters the one time password

In other word, I want to provide users the option to choose per account whether they want to enable MFA and if they do, I want them to have several options to choose from. This is very similar to how MongoDB Atlas is set up. They provide a multitude of MFA methods and on each login the user can choose. In many cases this provides the best user experience.

Would it be possible to add a hook to Accounts or the accounts-password package as such that developers have the option to set up their own custom login flow - a function that gets called during login. Accounts.validateLoginAttempt does allow running custom code during login, but that does not have the option to return anything to the client (aside from errors). For the above, if it is determined that an account has MFA enabled and no OTP is sent, then the login process needs to be stopped and a list of available OTP methods needs to be returned to the user. It is possible to use errors for this communication ([like here](https://blog.meteor.com/tutorial-two-factor-authentication-with-meteor-and-totp-21d4a2f9ee51)), but that seems like not a very good practice. Maybe validateLoginAttempt itself could be modified as such that it could return data to the client that's attempting the login? This would allow for a lot of flexibility, as developers can then introduce custom code to the whole login process.

On the other hand, my approach to security is usually to stick to default solutions as closely as possible and avoid writing custom code. Especially when it comes to authentication. So perhaps for security a better solution would be to just modify the accounts-password package to enable the use of multiple OTP methods simultaneously?

[Release 2.6.1](https://github.com/meteor/meteor/pull/11886) brings MFA support in accounts-password and accounts-passwordless, but it seems the above will not be possible with the update. Another option would be to use Accounts.registerLoginHandler and Accounts.callLoginMethod as described in this Meteor [blog post](https://blog.meteor.com/tutorial-two-factor-authentication-with-meteor-and-totp-21d4a2f9ee51). I did actually already implement such login in my app, but then decided to abandon that option. Relying on undocumented internal code for such a critical piece just seems wrong.

PS. I did say to Filipe a while back that I'd submit a PR to describe Accounts.registerLoginHandler and Accounts.callLoginMethod in the docs, but thinking about it, it probably would not be a good option. It would encourage people to write custom code on it and makes it more likely that in the future some change in Meteor core might lead to horrible security issues. A lot safer and cleaner option is to make the accounts- packages sufficiently flexible and not expose internal code.

PPS. It seemed best to create a separate dicussion thread for this in addition to [this](https://github.com/meteor/meteor/discussions/11514) discussion.

**Comment by denihs** [comment]
Hi @vooteles, 

With the new package `accounts-2fa`, I think that we'll be able to do this. As you can see in the [docs](https://deploy-preview-11886--meteor-docs.netlify.app/packages/accounts-2fa.html) of this package, it already has all the methods needed to create any 2FA system. But right now, it implements just the OTP one.

This [repo](https://github.com/denihs/test-2fa-package/tree/master) uses this package to implement the login flow. It lacks the second step you mentioned because it implements just the OTP one.

The package is just integrated with `accounts-password` and `accounts-passwordless` because it was an internal decision, but it's fairly simple to integrate it with other packages. In the docs, we even say how: https://deploy-preview-11886--meteor-docs.netlify.app/packages/accounts-2fa.html#integrating-auth-package.

I feel that a lot of you said we already got or would be easy to get with `accounts-2fa`, even though there is just OTP right now. 

Were you able to play around with it? I think it would be cool if you do and provide your feedback on how you feel about it and how far it's from your vision.

---

Also, I know you said it seemed best to create a separate discussion thread, but yours feels like  [this](https://github.com/meteor/meteor/discussions/11514) one... But I guess we can keep it like this for now.

**Reply by jkasevits** [reply]
Thanks for the quick reply! I think `accounts-2fa` provides a good base to build on and much of what is necessary already exists in it. It seems to me that in order to support more flexible login flows, including the one I described above, three additions would be needed to the package.

**Firstly, adding support for multiple ways to send a one-time-password to the user.** Currently the package supports one method, using an authenticator app. It would be great if there were an additional method under which upon login the server would generate a one-time-password and send it to the user. Ideally `accounts-2fa` would accept a configuration array on the server, something like this, so that developers can set up a multitude of ways that sending the one-time-password can happen:

```
[
  { name: 'authenticator-app' }, // current flow, nothing new is sent to the user, so no extra code is executed
  { name: 'sms', sendFunction: ... }, // developer provides code that calls a Twilio API
  { name: 'email', sendFunction: ... }, // developer provides code that calls a Postmark API
  { name: 'custom1', sendFunction: ... }, // developer provides code that calls some other custom API, for example a call to some connected app
]
```

Or maybe a cleaner way to do this would be to have something like Accounts.setUp2FAMethod(options) and if a developer wants to set up 2fa with both SMS and email, then it just calls it twice on startup, once per delivery method.

Regardless of the exact method of configuration, all these options would be 99% the same. Server generates the one-time-password, stores it in the user document and then executes some developer defined code to transport the one-time-password to the user.

**Secondly, modifying Accounts.has2faEnabled to return not just a boolean about whether 2fa is enabled, but also return a list which specific delivery methods are enabled for that specific user that is trying to log in.** So that on the login page we can show a list of 2fa options that the user can choose from, just like on the Mongo Atlas login page.

**Thirdly, creating something like Accounts.sendOTP(selector, deliveryMethod).** When on the login page a user clicks on SMS, this is called and the server sends out the one-time-password with the delivery method that the user chose.

Ideally, with these new 2fa methods, it would be possible to configure the complexity of the one-time-password as well in the server configuration. When using an authenticator app, that is already handled. But when users need to input the password themselves, then it might be useful to specify the lenght of the password and whether it is alphanumerical or just numbers etc.

How does the above sound? Would that be possible based on the current `accounts-2fa` package?
